### DISCUSSION TITLE
JTAG

### DISCUSSION THREAD

**Discussion by 0Grit** [esp-rs/esp-rust-board#3] [Ideas]
Please expose JTAG on a standard pin header in addition to the standard cortex debug 10 pin header.

**Comment by igrr** [comment]
Hi @0Grit,
I think we were considering to use the built-in USB-SERIAL-JTAG peripheral of ESP32-C3, as that would provide both JTAG and serial console while occupying only two pins of the chip, and requiring only one USB cable. This built-in adapter is now supported in probe-rs: https://github.com/probe-rs/probe-rs/pull/863.

Do you envision some situations where you would prefer to use an external probe instead of the one built into the chip?

**Reply by 0Grit** [reply]
If anybody wanted to create a fixture and do automated testing of the board they could. They would have complete control of the board while also being able to breakout the various buses for monitoring and test.

Highly recommend coming up with a standard pin header footprint to do this. Or extend an existing standard pin header footprint and declare the added pins specifically as jtag/debug pins.

I for example would like to have 5-10 of these boards in a rack and be able to do continuous integration tests and remote debug. Including monitoring I2C, spi etc.

**Reply by 0Grit** [reply]
@igrr We are genuinely excited about this board if it's done the right way.
Even if you break the signals out to an edge connector; don't care we'll make an adapter board in order to do fully automated on target testing and logging.

Here is an example of what I am talking about.
https://os.mbed.com/platforms/AGORA-DEV/
https://os.mbed.com/components/Flidor/

**Reply by pedrominatel** [reply]
@0Grit I see no problems with exposing the JTAG pins using the 10-pins 1.27mm.

Is there any special requirement for the connector? Usually, the 10-pin JTAG connector is more common these days.

**Reply by 0Grit** [reply]
Often its called "cortex debug connector".  It obviously needs a more generic name than that in this case.

Will there be any other connectors on this board?
I still want to be able to fixture this thing so I can do automated testing. I want to be able to breakout the buses etc. just by popping this board into a custom baseboard.

I'm thinking something like using the standard "feather" pinout but extend it a couple pins in order to support JTAG. And definitely bring that awesome usb/jtag perioheral over too.

I am already in the process of building out a cart specifically to do on target testing of Golioth's DSDK for other boards.

If you would like to do a call to discuss this stuff please do. I have a good bit of experience architecting boards like this. Happy to help any way I can.

**Reply by pedrominatel** [reply]
@0Grit the form factor of this dev board is still open. The idea is to use something already in the market (Adafruit Feather i.e) and not to create something new.
We can extend all pins to the 2,54mm headers, including the JTAG, and also add the 10pin JTAG connector.

We can make this call if you want and you'll be more than welcome to help on this project!

**Reply by 0Grit** [reply]
@pedrominatel As a professional who has both designed PCB's and specializes in firmware + IoT; I purposely avoid dev boards that use the feather footprint.

The feather footprint has no standard pins for debug. 
With no standard way to debug a feather board this also means there is no way to do automated tests generically across feather boards.

So what I am proposing is to use this as an opportunity to extend the feather footprint 'standard' in a way that is **backwards compatible**, that better accommodates professionals and possibly **production use cases as well**.

**Reply by pedrominatel** [reply]
This is a good idea. Let's think about this extended Feather footprint. If you have any suggestions for the pin layout, please share them with us.
